linode/terraform-provider-linode
# Add time format upgrade logic for stackscript resource

## 📝 Description

Implement time format upgrade logic to seamlessly upgrade the time format from something like`2024-04-05 21:42:22 +0000 UTC` to RFC3339, while preventing the error reported by RFC3339 format validation in the `timetypes.RFC3339Type` when the old format has been written into the state by a v1 Linode provider.

closes #1372 

## ✔️ How to Test

### Automated Testing

```bash
make PKG_NAME="linode/stackscript" unit-test
```

```bash
make PKG_NAME="linode/stackscript" int-test
```

### Manual Testing

- Turn off the [Terraform provider development overrides](https://developer.hashicorp.com/terraform/plugin/debugging#terraform-cli-development-overrides)
- Apply this Terraform config below with a v1 provider
```terraform
terraform {
  required_providers {
    linode = {
      source = "linode/linode"
      version = "1.30.0"
    }
  }
}
provider "linode" {
}

resource "linode_stackscript" "test" {
    label = "test"
    description = "Test stackscript"
    script = <<EOF
#!/bin/bash
echo "test"
EOF
    images = ["any/all"]
    rev_note = "v1"
}
```
- When changing provider version, you may see `Error: Inconsistent dependency lock file`. You can do `terraform init -upgrade` to update the lock file to fix it.
- After applying it with v1 provider, turn the [Terraform provider development overrides](https://developer.hashicorp.com/terraform/plugin/debugging#terraform-cli-development-overrides) back on.
- Do `go build` and then `terraform plan` with the provider in this branch to verify that no error is showing up.
